internal/plugin/external_plugins.go

```diff
@@ -0,0 +1,228 @@
+package plugin
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	go_plugin "plugin"
+	"strings"
+
+	"github.com/cyberark/secretless-broker/pkg/secretless/log"
+	"github.com/cyberark/secretless-broker/pkg/secretless/plugin/connector/http"
+	"github.com/cyberark/secretless-broker/pkg/secretless/plugin/connector/tcp"
+)
+
+// ExternalPluginLookupFunc returns all available external plugins.
+type ExternalPluginLookupFunc func(pluginDir string,
+	checksumfile string,
+	logger log.Logger,
+) (map[string]*go_plugin.Plugin, error)
+
+// LoadPluginsFromDir loads all plugins from a specified directory and returns
+// Plugins struct with tcp and http connectors.
+func LoadPluginsFromDir(pluginDir string,
+	checksumsFile string,
+	logger log.Logger,
+) (map[string]*go_plugin.Plugin, error) {
+
+	filenames, err := getCheckedPlugins(pluginDir, checksumsFile, logger)
+	if err != nil {
+		return nil, err
+	}
+
+	return loadPluginFiles(pluginDir, filenames, logger)
```

Comment:
Fixed but we're getting back the odd-looking `if`. Since VerifyPluginChecksums returns file objects it's either that or we have to make filenames array twice so I went with the former.

Reply:
_it is not the spoon that bends, it is only yourself..._

In this case you can either:

1. Change `VerifyPluginChecksums` to return full paths.  It's only being referenced in one place atm... my new code in entrypoint, but I don't even use the return value.  Downside here is you might have to update its tests.

2.  Create a quickie wrapper function, whose code you've actually already written.  I'd probably just make the function package private but it could even be local to this func.  Ie:
```go
	filePaths := func(path string, infos []os.FileInfo) {
		filePaths := []string{}
		for _, info := range infos {
			filePaths = append(filePaths, path.Join(path, info.Name()))
		}
	}
```
Then you can keep the old flat code:
```go
func getCheckedPlugins(
	pluginDir string,
	checksumsFile string,
	logger log.Logger,
) ([]*go_plugin.Plugin, error) {

	uncheckedFiles, err := ioutil.ReadDir(pluginDir)
	if err != nil {
		return nil, err
	}

	// no files found, return empty slice
	if len(uncheckedFiles) == 0 {
		return []*go_plugin.Plugin{}, nil
	}

	if checksumsFile == "" {
		logger.Warnln("Plugin hashes were not provided - tampering will not be detectable!")
		uncheckedPaths := filePaths(uncheckedFiles, pluginDir)
		return loadPluginFiles(uncheckedPaths, logger), nil
	}

	if checkedFiles, err = VerifyPluginChecksums(pluginDir, checksumsFile); err != nil {
		logger.Errorln(err)
		return nil, err
	}

	checkedPaths := filePaths(checkedFiles, pluginDir)
	return loadPluginFiles(checkedPaths, logger), nil
}
```

Reply:
> Change VerifyPluginChecksums to return full paths

This isn't advisable. While we don't use the functionality provided by this module, returned `os.FileInfo` has a higher guarantee that the underlying file object did not change between checking the checksum and loading the file. Even the move to compose the paths earlier is already reducing the assurance that the file we verified is the same one we are loading so weakening this further would not be good.

> Create a quickie wrapper function

I'll do this due to time but I disagree on the approach tbh if it's just to avoid a single + small if/else block but we can discuss this in slack at a later point.